drivers: display: GC9A01A: Add driver for GC9A01A

amrithvenkat opened the pull request:
Adding the driver for the GC9A01A 240x240 1.28 inch round LCD display.

amrithvenkat committed:
drivers: display: GC9A01A: Add driver for GC9A01A

Adding the driver for the GC9A01A 240x240 1.28 inch round LCD display.

Signed-off-by: Amrith Venkat Kesavamoorthi <amrith@mr-beam.org>

amrithvenkat committed:
drivers: display: GC9A01A: resolve compliance checks

Update yaml liniting and C formatting

Signed-off-by: Amrith Venkat Kesavamoorthi <amrith@mr-beam.org>

faxe1008 commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+/**
+ * Copyright (c) 2017 Jan Van Winkel <jan.van_winkel@dxplore.eu>
```
Don't copy other peoples copyright. Happened today to me too 
:^)

amrithvenkat replied:
Thanks for the heads up. will remove with the next commit.

faxe1008 commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+	struct spi_buf_set tx_bufs = {.buffers = &tx_buf, .count = 1U};
+
+	/* send command */
+	tx_buf.buf = &cmd;
```
Why not set those in the declaration also as you've done above for `tx_bufs`?

amrithvenkat replied:
changed

faxe1008 commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+	tx_buf.buf = &cmd;
+	tx_buf.len = 1U;
+
+	gpio_pin_set_dt(&config->cmd_data, GC9A01A_CMD);
```
Maybe also checking the return value here?

amrithvenkat replied:
added

faxe1008 commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+	}
+
+	data->pixel_format = pixel_format;
+	data->bytes_per_pixel = bytes_per_pixel;
```
You can use `BYTES_PER_PIXEL` here directly.

faxe1008 commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+ * @param dev GC9A01A device instance.
+ * @return 0 when successful, errno otherwise.
+ */
+static int gc9a01a_configure(const struct device *dev)
```
Maybe call it `_init` instead?

amrithvenkat replied:
I followed the same pattern as the ilitek drivers. There is already a function `gc9a01a_init()` and the` _configure` is called within the `_init`

faxe1008 commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+		pixel_format = PIXEL_FORMAT_RGB_888;
+	}
+
+	r = gc9a01a_set_pixel_format(dev, pixel_format); // Set pixel format.
```
Why not pass the cfg pixel format directly here and prevent bad values via `BUILD_CHECK`?

faxe1008 commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+	} else if (config->rotation == 180U) {
+		orientation = DISPLAY_ORIENTATION_ROTATED_180;
+	} else {
+		orientation = DISPLAY_ORIENTATION_ROTATED_270;
```
Same here. Also I guess a DT enum can be used to prevent bad values here, and if not again `BuILD_CHECK`.

amrithvenkat replied:
Could you point me to a resource or example. I am lost here by what you mean

amrithvenkat commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,599 @@
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->cmd_data.port)) {
```
will change in the next commit

amrithvenkat replied:
done

amrithvenkat committed:
drivers: display: GC9A01A: formatting

Remove C99 comments, add yamllint-ing. Remove other copyrights
Implement changes based on comments

Signed-off-by: Amrith Venkat Kesavamoorthi <amrith@mr-beam.org>

jfischer-no commented on drivers/display/Kconfig.gc9a01a:
```diff
@@ -0,0 +1,12 @@
+config GC9A01A
+	bool "GC9A01A display driver"
+	default y
+	depends on DT_HAS_WAVESHARE_GC9A01A_ENABLED
```
We do not implement drivers for displays or display resellers, but for display controller. Here, it cannot be `DT_HAS_WAVESHARE_GC9A01A_ENABLED`, maybe `DT_HAS_GALAXYCORE_GC9A01A_ENABLED`

jfischer-no commented on drivers/display/display_gc9a01a.h:
```diff
@@ -0,0 +1,189 @@
+#define GC9A01A_TEON 0x35     ///< Tearing effect line on
+#define GC9A01A_MADCTL 0x36   ///< Memory Access Control
+#define GC9A01A_VSCRSADD 0x37 ///< Vertical Scrolling Start Address
+#define GC9A01A_PIXFMT 0x3A   ///< COLMOD: Pixel Format Set
```
Usually when writing drivers, especially the first one, it is a good idea to use an existing one as a template. There would be copyright notices from the original or forgotten comments to change, but here `///< COLMOD: Pixel Format Set` looks a bit confusing. Searching for this pattern we find [Adafruit_GC9A01A.h](https://github.com/adafruit/Adafruit_GC9A01A/commit/763dc26bfe95276044fdb1b8c12a08cb0df9e1d9#diff-0e1829dee3da6a0668fb2a86074cd36e018f139b89dd906dfaa1445d87ba20e5)
![adafruit_gc9a01a](https://github.com/zephyrproject-rtos/zephyr/assets/9864546/ef95f4dd-2ff0-42f7-bcb0-26b383e83fca)
This is not acceptable.

jfischer-no replied:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#licensing It is not worth not having an Apache license for it as you can type these defines in relatively short time from the datasheet.

jfischer-no commented on drivers/display/display_gc9a01a.c:
```diff
@@ -0,0 +1,614 @@
+ * @param tx_len Length of buffer to be sent.
+ * @return 0 on success, errno otherwise.
+ */
+int gc9a01a_transmit(const struct device *dev, uint8_t cmd, const void *tx_data,
```
`static int gc9a01a_transmit(const struct device *dev, uint8_t cmd, const void *tx_data,`
There should not be any doxygen comments in this file.

jfischer-no commented:
https://github.com/zephyrproject-rtos/zephyr/pull/62934#discussion_r1334416979
Closing as not acceptable.

joelguittet commented:
@amrithvenkat @jfischer-no why stopping working on this driver ?

ljunquera commented:
There seems to be a lot of interest in this driver and it seems like it's close. Is there anybody who can help finish it off from the community?

joelguittet commented:
@ljunquera for sure, I m actually trying to have zephyr working on the esp32s3-touch-lch-1.28 from waveshare. 
I have difficulties with building lvgl and driver together, looks like it s more related to the integration with the board or bootloader, not sure....
The board has also accelerator/gyroscope that would be nice to integrate in zephyr too.
Will be happy to collaborate with other people on this, please ping if you are interested or currently working on this :-)

ljunquera commented:
@joelguittet I would love to help but am a real novice. I am fighting though some of the cmake stuff right now. I know @jakkra has it working in ZSWatch. However I can help I will.

amrithvenkat commented:
@joelguittet @ljunquera I would like for the driver to be included in the code base. Even @jakkra has the implementation which is also similar to my PR. I face licensing issues which I'll be glad If anyone helps me out to resolve. Thanks

joelguittet commented:
@jakkra implementation need few improvement as far as I have seen. For example it is limiting one display, instead of foreach each DTS entry.
By the way I agree this is a good starting point, I m actually using his driver (or at least I try!)
So far I m able to build simple app for esp32s3 but as soon as I'm adding some specific lvgl stuff (and particularly the function lv_scr_act) the binary becomes really bigger (60kB) and the application doesn't boot at all. The device remains stuck at the end of the bootloader.
@ljunquera @amrithvenkat do you have the esp32s3 touch LCD hardware from waveshare or something else ?

ljunquera commented:
I'm trying to build it for the nRF5340 DK using the lvgl sample as the base. I have the [Waveshare 1.28inch touch LCD](https://www.waveshare.com/wiki/1.28inch_Touch_LCD) and the nRF Connect for VS Add-In. My current problems are basic (CMakeLists and overlay node)...again I'm new to this.

amrithvenkat commented:
@ljunquera I used the same sample from waveshare as the reference. And I was able to use multiple displays with my implementation.
@joelguittet implementation I did was on the Rpi pico rp2040. Unfortunately I don't have sample if it works with esp or nrf. Even ZSWatch implementation is on rp2040.

ljunquera commented:
Thanks, @amrithvenkat. I will try to get it working with nRF.

joelguittet commented:
Great example @amrithvenkat 
I have not rpi pico but will try as well, on the esp32s3.
I see you have not lvgl, looks a good step first to test the display.

About licensing issues I think the main solution is that the driver should be in "zephyr style". It s true we can't just copy from another source, but of course the goal is not to reinvent everything, it's possible to get inspiration from other examples from Arduino, esp IDF etc
My approach will be more or less to copy an existing and similar driver from zephyr, and modify where it is required.

ljunquera commented:
@joelguittet, what's a good driver to start with which is already in Zephyr?

joelguittet commented:
@ljunquera the one with spi bus, command/data and reset gpio will be a good starting point. St7789 for example?
This has also the advantage to not create new convention, for example cd and rst gpio should be called with the same name, in DTS and in the source too.

MrMarteng commented:
Hi guys,
I also started working on this driver as part of support for m5stack m5dial. I didn't have access to the initial branch therefore I started reworking requests on a separate branch: 
https://github.com/MrMarteng/zephyr/tree/gc9a91a_display

ljunquera commented:
Thanks, @MrMarteng. Great work. Are we going to run into the same copyright issues which caused a problem in the first pass?

MrMarteng commented:
> Thanks, @MrMarteng. Great work. Are we going to run into the same copyright issues which caused a problem in the first pass? Should we start with what @joelguittet recommended and start with a new build based on the existing st7789 and a copyright consistent with Zephyr Project.

Yes I saw the issue. I recreated the according files from scratch. Give me a few hours, I will push the latest version to the mentioned branch.

joelguittet commented:
@MrMarteng awesome! Will check it soon

ljunquera commented:
It would be great to have a sample to go with it.

amrithvenkat commented:
Work on this driver continued on PR - #67012
